scripts/gitlab_runner_pre_build/pre_build.py

```diff
@@ -0,0 +1,66 @@
+"""
+This script is responsible for taking the ROLE_ARN and GITLAB_OIDC_TOKEN from the environment and
+exchange it for a temporary AWS credentials. These credentials are then printed to stdout to be
+sourced by the gitlab runner in its pre_build configuration option.
+
+In the case of a PR build, the temporary credentials are scoped down to only allow access to the
+S3 bucket prefix for the relevant PR.
+"""
+import boto3, os, sys, jwt, json
+
+TEMPORARY_CREDENTIALS_DURATION = 3600 * 6  # 6 hours
+
+if __name__ == "__main__":
+    if not os.environ.get("ROLE_ARN") or not os.environ.get("GITLAB_OIDC_TOKEN"):
+        print("ROLE_ARN and GITLAB_OIDC_TOKEN not in the environment", file=sys.stderr)
+        sys.exit(0)  # this isn't an error yet.
+
+    # get aws sts temporary credentials
+    sts = boto3.client("sts")
+    assume_role_kwargs = {
+        "RoleArn": os.environ["ROLE_ARN"],
+        "RoleSessionName": (
+            f'GitLabRunner-{os.environ["CI_JOB_ID"]}-{os.environ["CI_COMMIT_SHORT_SHA"]}'
+        ),
+        "WebIdentityToken": os.environ["GITLAB_OIDC_TOKEN"],
+        "DurationSeconds": TEMPORARY_CREDENTIALS_DURATION,
+    }
+
+    print(
+        f"Assuming role {assume_role_kwargs['RoleArn']} with session name {assume_role_kwargs['RoleSessionName']}",
+        file=sys.stderr,
+    )
+
+    # if this is a PR build, narrow down the permissions to only allow access to the PR build prefix
+    token = jwt.decode(
+        os.environ["GITLAB_OIDC_TOKEN"],
+        algorithms=["RS256"],
+        # it's unnecessary to verify the signature here since this is running before any
+        # untrusted code. it's also useful to avoid having to install the crypto libraries
+        # in diverse environments.
+        options={"verify_signature": False},
+    )
+
+    if token["aud"] == "pr_binary_mirror":
```

Comment:
I see how you set this up in the gitlab job variables now, in the spack [PR](https://github.com/spack/spack/pull/39813).  This seems like a great benefit of this approach :+1: